pages/funnel-analytics/mock-funnel.ts

```diff
@@ -0,0 +1,129 @@
+  },
+
+  funnelError(props): void {
+    funnelMetricsLog.push({ action: 'funnelError', props });
+  },
+
+  funnelComplete(props): void {
+    funnelMetricsLog.push({ action: 'funnelComplete', props });
+  },
+
+  funnelSuccessful(props): void {
+    funnelMetricsLog.push({ action: 'funnelSuccessful', props });
+  },
+
+  funnelCancelled(props): void {
+    funnelMetricsLog.push({ action: 'funnelCancelled', props });
+  },
+
+  funnelChange(props): void {
+    funnelMetricsLog.push({ action: 'funnelChange', props });
+  },
+
+  funnelStepStart(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    funnelMetricsLog.push({ action: 'funnelStepStart', props, resolvedProps: { stepName } });
+  },
+
+  funnelStepComplete(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    funnelMetricsLog.push({ action: 'funnelStepComplete', props, resolvedProps: { stepName } });
+  },
+
+  funnelStepNavigation(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    // const subStepAllElements = document.querySelectorAll(props.subStepAllSelector); // TODO: Does not work
+
+    funnelMetricsLog.push({ action: 'funnelStepNavigation', props, resolvedProps: { stepName } });
+  },
+
+  funnelStepError(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    funnelMetricsLog.push({ action: 'funnelStepError', props, resolvedProps: { stepName } });
+  },
+
+  funnelStepChange(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    funnelMetricsLog.push({ action: 'funnelStepChange', props, resolvedProps: { stepName } });
+  },
+
+  funnelSubStepStart(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
+    const subStepName = document.querySelector(props.subStepNameSelector)!.innerHTML;
+    const subStepAllElements = document.querySelectorAll(props.subStepAllSelector);
+    const subStepElement = document.querySelector(props.subStepSelector);
+
+    funnelMetricsLog.push({
+      action: 'funnelSubStepStart',
+      props,
+      resolvedProps: { stepName, subStepName, subStepAllElements, subStepElement },
+    });
+  },
+
+  funnelSubStepComplete(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)?.innerHTML;
+    const subStepName = document.querySelector(props.subStepNameSelector)?.innerHTML;
+    const subStepAllElements = document.querySelectorAll(props.subStepAllSelector);
+    const subStepElement = document.querySelector(props.subStepSelector);
+
+    funnelMetricsLog.push({
+      action: 'funnelSubStepComplete',
+      props,
+      resolvedProps: { stepName, subStepName, subStepAllElements, subStepElement },
+    });
+  },
+
+  funnelSubStepError(props): void {
+    const stepName = document.querySelector(props.stepNameSelector)!.innerHTML;
```

Comment:
In some places you've used `?` and in some places you've used `!`. I'd recommend to use `?` everywhere, so that we still have the event even when some data is missing